File: api/spp_extras/spp_extras_api/helpers/quests.py
def all_completed_quests(chars, reg, weekly):
    all = {
      'alliance': {},
      'horde': {}
    }
    
    def addQuest(quest, type):
        guid = str(quest['guid'])
        questId = str(quest['quest'])
        faction = chars[guid]

        if not hasattr(all[faction], guid):
            all[faction][guid] = { 'reg': {}, 'weekly': {} }

        if type == 'reg':
            all[faction][guid]['reg'][questId] = quest
        elif type == 'weekly':
            all[faction][guid]['weekly'][questId] = quest

    for q in reg: addQuest(q, 'reg')
    for q in weekly: addQuest(q, 'weekly')
    print(all)
    return all


def all_quests(quests):
    all = {
        'alliance': {},
        'horde': {},
        'both': {}
    }

    alliance = [1, 4, 5, 8, 64, 65, 68, 77, 1101]
    horde = [2, 16, 32, 128, 130, 144, 162, 178, 690]
    both = [0, 255]

    for quest in quests:
        required_races = quest['requiredraces']
        entry = quest['entry']
        if required_races in alliance: all['alliance'][entry] = quest
        elif required_races in horde: all['horde'][entry] = quest
        elif required_races in both: all['both'][entry] = quest
        
    return all

File: client/types/quests.ts
import { Faction } from './factions';


export interface AllQuests {
  completedQuests: CompletedQuests | Record<string,never>;
  templateQuests: TemplateQuests | Record<string,never>;
}

export interface CharacterQuests {
  regular: Quests;
  daily?: Quests;
  weekly: Quests;
  monthly?: Quests;
}

export type CharacterQuestClass = 1 | 2 | 4 | 8 | 16 | 32 | 64 | 128 | 256 | 1024;

export type CharacterQuestRace = 1 | 2 | 4 | 8 | 16 | 32 | 64 | 128 | 512 | 1024;

export interface CompletedQuests {
  alliance: FactionQuests;
  horde: FactionQuests;
}

export interface FactionQuests {
  [key: string]: CharacterQuests;
}

export interface Quest {
  guid: number;
  quest: number;
  status?: number;
  rewarded?: number;
  explored?: number;
  timer?: number;
  mobcount1?: number;
  mobcount2?: number;
  mobcount3?: number;
  mobcount4?: number;
  itemcount1?: number;
  itemcount2?: number;
  itemcount3?: number;
  itemcount4?: number;
}

export interface QuestFlags {
  regular: number[];
  daily?: number[];
  weekly: number[];
  monthly?: number[];
}

export interface QuestRace {
  questRaceId: number;
  faction: Faction;
  races: string;
  raceIds: number[];
}

export interface QuestRaces {
  [key: string]: QuestRace;
}

export interface Quests {
  [key: string]: Quest;
}

export type QuestType = 'regular' | 'daily' | 'weekly' | 'monthly';

export interface TemplateFactionQuests {
  [key: string]: TemplateQuest;
}

export interface TemplateQuest {
  entry: number;
  requiredclasses: number;
  requiredraces: number;
  title: string;
  type: number;
  zoneorsort: number;
  questflags: number;
}

export interface TemplateQuests {
  alliance: TemplateFactionQuests | Record<string,never>;
  horde: TemplateFactionQuests | Record<string,never>;
  both: TemplateFactionQuests | Record<string,never>;
}
